File: backend/app/schemas/npo.py
"""Pydantic schemas for NPO (Non-Profit Organization) management."""

import uuid
from datetime import datetime

from pydantic import BaseModel, EmailStr, Field, HttpUrl, field_validator

from app.models.npo import NPOStatus
from app.schemas.npo_branding import BrandingResponse

# ================================
# Request Schemas
# ================================


class NPOCreateRequest(BaseModel):
    """Request schema for creating a new NPO (draft or application)."""

    name: str = Field(..., min_length=2, max_length=255)
    tagline: str | None = Field(default=None, max_length=255)
    description: str | None = Field(default=None, max_length=5000)
    mission_statement: str | None = Field(default=None, max_length=5000)
    tax_id: str | None = Field(default=None, max_length=50)
    website_url: HttpUrl | None = Field(default=None)
    phone: str | None = Field(default=None, max_length=20)
    email: EmailStr = Field(...)
    address: dict[str, str] | None = Field(
        default=None,
        description="Address JSON: {street, street2, city, state, postal_code, country}",
    )
    registration_number: str | None = Field(default=None, max_length=100)

    @field_validator("website_url")
    @classmethod
    def convert_url_to_string(cls, v: HttpUrl | None) -> str | None:
        """Convert HttpUrl to string for database storage."""
        return str(v) if v else None

    @field_validator("email")
    @classmethod
    def email_must_be_lowercase(cls, v: str) -> str:
        """Ensure email is lowercase."""
        return v.lower()

    @field_validator("name")
    @classmethod
    def name_must_not_be_empty(cls, v: str) -> str:
        """Ensure name is not just whitespace."""
        if not v.strip():
            raise ValueError("NPO name cannot be empty or whitespace")
        return v.strip()


class NPOUpdateRequest(BaseModel):
    """Request schema for updating NPO details."""

    name: str | None = Field(default=None, min_length=2, max_length=255)
    tagline: str | None = Field(default=None, max_length=255)
    description: str | None = Field(default=None, max_length=5000)
    mission_statement: str | None = Field(default=None, max_length=5000)
    tax_id: str | None = Field(default=None, max_length=50)
    website_url: HttpUrl | None = Field(default=None)
    phone: str | None = Field(default=None, max_length=20)
    email: EmailStr | None = Field(default=None)
    address: dict[str, str] | None = Field(default=None)
    registration_number: str | None = Field(default=None, max_length=100)

    @field_validator("website_url")
    @classmethod
    def convert_url_to_string(cls, v: HttpUrl | None) -> str | None:
        """Convert HttpUrl to string for database storage."""
        return str(v) if v else None

    @field_validator("email")
    @classmethod
    def email_must_be_lowercase(cls, v: str | None) -> str | None:
        """Ensure email is lowercase."""
        return v.lower() if v else None

    @field_validator("name")
    @classmethod
    def name_must_not_be_empty(cls, v: str | None) -> str | None:
        """Ensure name is not just whitespace."""
        if v is not None and not v.strip():
            raise ValueError("NPO name cannot be empty or whitespace")
        return v.strip() if v else None


class NPOStatusUpdateRequest(BaseModel):
    """Request schema for updating NPO status (admin only)."""

    status: NPOStatus
    notes: str | None = Field(None, max_length=5000, description="Admin notes for status change")


class NPOListRequest(BaseModel):
    """Request schema for listing NPOs with filters."""

    status: NPOStatus | None = Field(default=None)
    search: str | None = Field(default=None, max_length=255, description="Search by name or email")
    created_by_user_id: uuid.UUID | None = Field(default=None)
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)


# ================================
# Response Schemas
# ================================


class NPOResponse(BaseModel):
    """Response schema for NPO details."""

    id: uuid.UUID
    name: str
    tagline: str | None
    description: str | None
    mission_statement: str | None
    tax_id: str | None
    website_url: str | None
    phone: str | None
    email: str
    address: dict[str, str] | None
    registration_number: str | None
    status: NPOStatus
    created_by_user_id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    # Related data counts (optional, for list views)
    member_count: int | None = None
    active_member_count: int | None = None

    # Branding info (optional, for selectors/list views)
    logo_url: str | None = None

    model_config = {"from_attributes": True}


class NPODetailResponse(NPOResponse):
    """Detailed response schema for NPO with related entities."""

    # Include branding info
    branding: BrandingResponse | None = Field(None, description="NPO branding information")
    # Include application info (if exists)
    application: dict[str, str] | None = Field(None, description="NPO application details")


class NPOListResponse(BaseModel):
    """Response schema for paginated NPO list."""

    items: list[NPOResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class NPOCreateResponse(BaseModel):
    """Response schema after creating NPO."""

    npo: NPOResponse
    message: str = "NPO created successfully"

File: frontend/fundrbolt-admin/src/types/npo.ts
/**
 * NPO (Non-Profit Organization) types
 * Type definitions for NPO management, applications, members, and branding
 */

// ============================================
// Enums
// ============================================

export type NPOStatus = 'draft' | 'pending_approval' | 'approved' | 'suspended' | 'rejected'

export type ApplicationStatus = 'submitted' | 'under_review' | 'approved' | 'rejected'

export type MemberRole = 'admin' | 'co_admin' | 'staff'

export type MemberStatus = 'active' | 'invited' | 'suspended' | 'removed'

export type InvitationStatus = 'pending' | 'accepted' | 'expired' | 'revoked'

// ============================================
// NPO Types
// ============================================

export interface NPO {
  id: string
  name: string
  tagline: string | null
  description: string | null
  mission_statement: string | null
  tax_id: string | null
  website_url: string | null
  phone: string | null
  email: string
  address: {
    street?: string
    street2?: string
    city?: string
    state?: string
    postal_code?: string
    country?: string
  } | null
  registration_number: string | null
  status: NPOStatus
  created_by_user_id: string
  created_at: string
  updated_at: string
  member_count?: number
  active_member_count?: number
}

export interface NPODetail extends NPO {
  branding?: NPOBranding
  application?: NPOApplication
}

export interface NPOCreateRequest {
  name: string
  tagline?: string
  description?: string
  mission_statement?: string
  tax_id?: string
  website_url?: string
  phone?: string
  email: string
  address?: {
    street?: string
    street2?: string
    city?: string
    state?: string
    postal_code?: string
    country?: string
  }
  registration_number?: string
}

export interface NPOUpdateRequest {
  name?: string
  tagline?: string
  description?: string
  mission_statement?: string
  tax_id?: string
  website_url?: string
  phone?: string
  email?: string
  address?: {
    street?: string
    street2?: string
    city?: string
    state?: string
    postal_code?: string
    country?: string
  }
  registration_number?: string
}

export interface NPOListParams {
  status?: NPOStatus
  search?: string
  created_by_user_id?: string
  page?: number
  page_size?: number
}

export interface NPOListResponse {
  items: NPO[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

// ============================================
// NPO Application Types
// ============================================

export interface NPOApplication {
  id: string
  npo_id: string
  status: ApplicationStatus
  review_notes: Record<string, string> | null
  reviewed_by_user_id: string | null
  submitted_at: string
  reviewed_at: string | null
  created_at: string
  updated_at: string
  npo_name?: string
  npo_email?: string
}

export interface ApplicationReviewRequest {
  status: ApplicationStatus
  review_notes?: Record<string, string>
}

export interface ApplicationListParams {
  status?: ApplicationStatus
  npo_id?: string
  page?: number
  page_size?: number
}

export interface ApplicationListResponse {
  items: NPOApplication[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

// ============================================
// NPO Member Types
// ============================================

export interface NPOMember {
  id: string
  npo_id: string
  user_id: string
  role: MemberRole
  status: MemberStatus
  joined_at: string | null
  invited_by_user_id: string | null
  created_at: string
  updated_at: string
  user_email?: string
  user_first_name?: string
  user_last_name?: string
  user_full_name?: string
  invited_by_name?: string
}

export interface MemberInviteRequest {
  email: string
  role: MemberRole
  first_name?: string
  last_name?: string
}

export interface MemberAddRequest {
  user_id: string
  role: MemberRole
}

export interface MemberRoleUpdateRequest {
  role: MemberRole
}

export interface MemberStatusUpdateRequest {
  status: MemberStatus
  notes?: string
}

export interface MemberListParams {
  npo_id?: string
  status?: MemberStatus
  role?: MemberRole
  search?: string
  page?: number
  page_size?: number
}

export interface MemberListResponse {
  items: NPOMember[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface MemberInviteResponse {
  invitation_id: string
  email: string
  role: MemberRole
  expires_at: string
  message: string
}

export interface PendingInvitation {
  id: string
  email: string
  role: MemberRole
  expires_at: string
  created_at: string
}

// ============================================
// NPO Branding Types
// ============================================

export interface NPOBranding {
  id: string
  npo_id: string
  primary_color: string | null
  secondary_color: string | null
  background_color: string | null
  accent_color: string | null
  logo_url: string | null
  social_media_links: {
    facebook?: string
    twitter?: string
    instagram?: string
    linkedin?: string
  } | null
  custom_css_properties: Record<string, string> | null
  created_at: string
  updated_at: string
  npo_name?: string
}

export interface BrandingCreateRequest {
  primary_color?: string
  secondary_color?: string
  background_color?: string
  accent_color?: string
  logo_url?: string
  social_media_links?: {
    facebook?: string
    twitter?: string
    instagram?: string
    linkedin?: string
  }
  custom_css_properties?: Record<string, string>
}

export interface BrandingUpdateRequest {
  primary_color?: string
  secondary_color?: string
  background_color?: string
  accent_color?: string
  logo_url?: string
  social_media_links?: {
    facebook?: string
    twitter?: string
    instagram?: string
    linkedin?: string
  }
  custom_css_properties?: Record<string, string>
}

export interface LogoUploadRequest {
  file_name: string
  file_size: number
  content_type: string
}

export interface LogoUploadResponse {
  upload_url: string
  logo_url: string
  expires_in: number
  message: string
}
